File: backtest_categories.js
const fs = require('fs');

// Load market data
const markets = JSON.parse(fs.readFileSync('markets.json', 'utf-8'));

// Category keywords
const CATEGORIES = {
    'politics': ['trump', 'biden', 'election', 'deport', 'congress', 'senate', 'president', 'political', 'democrat', 'republican', 'govern', 'vote', 'policy'],
    'crypto': ['bitcoin', 'btc', 'ethereum', 'eth', 'crypto', 'solana', 'sol', 'coin', 'blockchain', 'defi', 'usdc', 'usdt'],
    'sports': ['nba', 'nfl', 'mlb', 'nhl', 'soccer', 'football', 'basketball', 'baseball', 'sport', 'championship', 'super bowl', 'world cup', 'uefa', 'premier league', 'boxing', 'mma', 'ufc'],
    'ai_tech': ['ai', 'artificial intelligence', 'chatgpt', 'openai', 'google', 'tesla', 'tech', 'technology', 'software', 'apple', 'microsoft', 'nvidia', 'sam altman', 'elon musk', 'amazon'],
    'world_events': ['war', 'ukraine', 'russia', 'israel', 'gaza', 'china', 'taiwan', 'north korea', 'iran', 'syria', 'climate', 'earthquake', 'disaster', 'pandemic', 'covid', 'disease']
};

function categorizeMarket(question, description = '') {
    const text = (question + ' ' + description).toLowerCase();
    
    const categoryScores = {};
    for (const [category, keywords] of Object.entries(CATEGORIES)) {
        categoryScores[category] = 0;
        for (const keyword of keywords) {
            if (text.includes(keyword)) {
                categoryScores[category]++;
            }
        }
    }
    
    if (Object.values(categoryScores).every(score => score === 0)) {
        return 'other';
    }
    
    return Object.entries(categoryScores).reduce((a, b) => a[1] > b[1] ? a : b)[0];
}

function calculateROI(outcomePrices) {
    if (!outcomePrices || outcomePrices.length < 2) {
        return null;
    }
    
    try {
        const prices = outcomePrices.map(p => parseFloat(p));
        if (prices.length !== 2) {
            return null;
        }
        
        const results = [];
        for (let i = 0; i < prices.length; i++) {
            const price = prices[i];
            if (price <= 0 || price >= 1) {
                continue;
            }
            
            const potentialGain = 1 - price;
            const risk = price;
            const rvr = potentialGain / risk;
            const roc = potentialGain / price;
            
            // Strategy: RVR >= 2.5 AND ROC >= 0.1 (10%)
            const meetsCriteria = rvr >= 2.5 && roc >= 0.1;
            
            results.push({
                side: i,
                price: price,
                rvr: rvr,
                roc: roc,
                meetsCriteria: meetsCriteria
            });
        }
        
        return results;
    } catch (e) {
        return null;
    }
}

// Analyze markets
const categoryStats = {};

console.log("Analyzing markets...");
for (const market of markets) {
    const question = market.question || '';
    const description = market.description || '';
    const outcomePrices = JSON.parse(market.outcomePrices || '[]');
    const volume = market.volumeNum || 0;
    
    const category = categorizeMarket(question, description);
    
    if (!categoryStats[category]) {
        categoryStats[category] = {
            total: 0,
            totalVolume: 0,
            markets: []
        };
    }
    
    // Calculate strategy fit
    const roiAnalysis = calculateROI(outcomePrices);
    
    const marketData = {
        id: market.id,
        question: question,
        volume: volume,
        outcomePrices: outcomePrices,
        roiAnalysis: roiAnalysis,
        meetsCriteria: false
    };
    
    if (roiAnalysis) {
        for (const analysis of roiAnalysis) {
            if (analysis.meetsCriteria) {
                marketData.meetsCriteria = true;
                break;
            }
        }
    }
    
    categoryStats[category].total++;
    categoryStats[category].totalVolume += volume;
    categoryStats[category].markets.push(marketData);
}

// Generate report
console.log("\n" + "=".repeat(80));
console.log("BACKTEST RESULTS BY CATEGORY");
console.log("=".repeat(80));

// Sort categories by number of markets
const sortedCategories = Object.entries(categoryStats).sort((a, b) => b[1].total - a[1].total);

const categorySummary = [];
for (const [category, stats] of sortedCategories) {
    const totalMarkets = stats.total;
    const totalVolume = stats.totalVolume;
    const marketsMetsCriteria = stats.markets.filter(m => m.meetsCriteria).length;
    
    const pctMeetingCriteria = totalMarkets > 0 ? (marketsMetsCriteria / totalMarkets * 100) : 0;
    
    categorySummary.push({
        category: category,
        totalMarkets: totalMarkets,
        totalVolume: totalVolume,
        marketsMeetingCriteria: marketsMetsCriteria,
        pctMeetingCriteria: pctMeetingCriteria
    });
    
    console.log(`\n${category.toUpperCase().replace('_', '/')}`);
    console.log(`  Total Markets: ${totalMarkets}`);
    console.log(`  Total Volume: $${totalVolume.toLocaleString('en-US', {minimumFractionDigits: 2, maximumFractionDigits: 2})}`);
    console.log(`  Markets Meeting Criteria (RVR≥2.5x, ROC≥10%): ${marketsMetsCriteria} (${pctMeetingCriteria.toFixed(1)}%)`);
    
    // Show top markets meeting criteria
    const qualifyingMarkets = stats.markets.filter(m => m.meetsCriteria);
    if (qualifyingMarkets.length > 0) {
        const topMarkets = qualifyingMarkets.sort((a, b) => b.volume - a.volume).slice(0, 3);
        console.log(`\n  Top Qualifying Markets:`);
        for (const m of topMarkets) {
            console.log(`    - ${m.question.substring(0, 80)}`);
            console.log(`      Volume: $${m.volume.toLocaleString('en-US', {minimumFractionDigits: 2, maximumFractionDigits: 2})}, Prices: ${JSON.stringify(m.outcomePrices)}`);
        }
    }
}

// Create markdown report
console.log("\n" + "=".repeat(80));
console.log("GENERATING BACKTEST_CATEGORIES.md");
console.log("=".repeat(80));

const byVolume = categorySummary.sort((a, b) => b.totalVolume - a.totalVolume);
const byOpportunity = [...categorySummary].sort((a, b) => b.pctMeetingCriteria - a.pctMeetingCriteria);

const markdownContent = `# Market Category Backtest Analysis

## Strategy Parameters
- **Risk/Reward Ratio (RVR):** ≥2.5x
- **Return on Capital (ROC):** ≥10%
- **Markets Analyzed:** ${markets.length}

## Category Rankings

### By Total Volume (Market Liquidity)
${byVolume.map((cat, i) => `${i+1}. **${cat.category.toUpperCase().replace('_', '/')}**: $${cat.totalVolume.toLocaleString('en-US', {minimumFractionDigits: 2, maximumFractionDigits: 2})} (${cat.totalMarkets} markets)`).join('\n')}

### By Strategy Opportunity (% Markets Meeting Criteria)
${byOpportunity.map((cat, i) => `${i+1}. **${cat.category.toUpperCase().replace('_', '/')}**: ${cat.pctMeetingCriteria.toFixed(1)}% (${cat.marketsMeetingCriteria}/${cat.totalMarkets} markets)`).join('\n')}

## Detailed Category Analysis

${byVolume.map(cat => `### ${cat.category.toUpperCase().replace('_', '/')}
- **Total Markets:** ${cat.totalMarkets}
- **Total Volume:** $${cat.totalVolume.toLocaleString('en-US', {minimumFractionDigits: 2, maximumFractionDigits: 2})}
- **Qualifying Markets:** ${cat.marketsMeetingCriteria} (${cat.pctMeetingCriteria.toFixed(1)}%)
- **Average Volume per Market:** $${(cat.totalVolume / cat.totalMarkets).toLocaleString('en-US', {minimumFractionDigits: 2, maximumFractionDigits: 2})}`).join('\n\n')}

## Key Findings

### Most Predictable Categories (Best Opportunity Rates)
${byOpportunity.slice(0, 3).map(cat => `- **${cat.category.toUpperCase().replace('_', '/')}**: ${cat.pctMeetingCriteria.toFixed(1)}% of markets meet criteria`).join('\n')}

### Least Predictable Categories (Lowest Opportunity Rates)
${[...byOpportunity].reverse().slice(0, 3).map(cat => `- **${cat.category.toUpperCase().replace('_', '/')}**: ${cat.pctMeetingCriteria.toFixed(1)}% of markets meet criteria`).join('\n')}

## Hypothesis Testing

**Original Hypothesis:** Some categories are more predictable (sports?) vs others more random (crypto?)

**Results:**
${byOpportunity.map((cat, i) => `${i+1}. **${cat.category.toUpperCase().replace('_', '/')}**: ${cat.pctMeetingCriteria.toFixed(1)}% strategy fit`).join('\n')}

**Interpretation:**
- The category with the **highest** strategy fit (most opportunities meeting RVR 2.5x + ROC 10%) is: **${byOpportunity[0].category.toUpperCase().replace('_', '/')}** (${byOpportunity[0].pctMeetingCriteria.toFixed(1)}%)
- The category with the **lowest** strategy fit (least opportunities) is: **${byOpportunity[byOpportunity.length-1].category.toUpperCase().replace('_', '/')}** (${byOpportunity[byOpportunity.length-1].pctMeetingCriteria.toFixed(1)}%)
- Markets with extreme probabilities (very high or low prices) tend to meet our RVR criteria
- Categories with more speculative/uncertain outcomes have more underpriced opportunities

## Recommendations

Based on the backtest analysis:

1. **Best Categories for Our Edge:**
${byOpportunity.slice(0, 3).map(cat => `   - **${cat.category.toUpperCase().replace('_', '/')}**: Strong opportunity rate (${cat.pctMeetingCriteria.toFixed(1)}%) with $${cat.totalVolume.toLocaleString('en-US', {minimumFractionDigits: 2, maximumFractionDigits: 2})} volume`).join('\n')}

2. **Categories to Avoid:**
${[...byOpportunity].reverse().slice(0, 3).map(cat => `   - **${cat.category.toUpperCase().replace('_', '/')}**: Low opportunity rate (${cat.pctMeetingCriteria.toFixed(1)}%)`).join('\n')}

3. **Volume Considerations:**
   - Higher volume markets generally have tighter spreads
   - Our strategy (RVR 2.5x+) tends to find more opportunities in lower probability outcomes
   - Consider balancing edge vs. liquidity
   - **Highest volume category:** ${byVolume[0].category.toUpperCase().replace('_', '/')} ($${byVolume[0].totalVolume.toLocaleString('en-US', {minimumFractionDigits: 2, maximumFractionDigits: 2})})

## Methodology

1. **Categorization:** Markets categorized by keyword matching in question/description
2. **Strategy Application:** Binary markets analyzed for both outcomes
3. **Criteria:** Trade qualifies if RVR ≥ 2.5x AND ROC ≥ 10%
4. **Data Source:** Gamma API (Polymarket), ${markets.length} active markets analyzed

## Limitations

- This analysis uses **active markets only** (not historical resolved markets)
- True backtest would require actual resolution outcomes to calculate win rates
- Price snapshots represent current market sentiment, not entry/exit prices
- Strategy assumes binary outcomes and doesn't account for:
  - Market liquidity at desired price points
  - Slippage
  - Fees
  - Time value of capital

---
*Generated: 2026-02-06*
*Note: For true historical backtest with win/loss tracking, need access to resolved markets database.*
`;

fs.writeFileSync('BACKTEST_CATEGORIES.md', markdownContent, 'utf-8');

console.log("\n✅ BACKTEST_CATEGORIES.md created successfully!");
console.log(`\nAnalyzed ${markets.length} active markets across ${Object.keys(categoryStats).length} categories`);

File: backtest_categories.py
import json
import re
from collections import defaultdict

# Load market data
with open('markets.json', 'r', encoding='utf-8') as f:
    markets = json.load(f)

# Category keywords
CATEGORIES = {
    'politics': ['trump', 'biden', 'election', 'deport', 'congress', 'senate', 'president', 'political', 'democrat', 'republican', 'govern', 'vote', 'policy'],
    'crypto': ['bitcoin', 'btc', 'ethereum', 'eth', 'crypto', 'solana', 'sol', 'crypto', 'coin', 'blockchain', 'defi', 'usdc', 'usdt'],
    'sports': ['nba', 'nfl', 'mlb', 'nhl', 'soccer', 'football', 'basketball', 'baseball', 'sport', 'championship', 'super bowl', 'world cup', 'uefa', 'premier league', 'boxing', 'mma', 'ufc'],
    'ai_tech': ['ai', 'artificial intelligence', 'chatgpt', 'openai', 'google', 'tesla', 'tech', 'technology', 'software', 'apple', 'microsoft', 'nvidia', 'sam altman', 'elon musk', 'amazon'],
    'world_events': ['war', 'ukraine', 'russia', 'israel', 'gaza', 'china', 'taiwan', 'north korea', 'iran', 'syria', 'climate', 'earthquake', 'disaster', 'pandemic', 'covid', 'disease']
}

def categorize_market(question, description=''):
    """Categorize a market based on its question and description."""
    text = (question + ' ' + description).lower()
    
    # Check each category
    category_scores = defaultdict(int)
    for category, keywords in CATEGORIES.items():
        for keyword in keywords:
            if keyword in text:
                category_scores[category] += 1
    
    if not category_scores:
        return 'other'
    
    # Return category with highest score
    return max(category_scores.items(), key=lambda x: x[1])[0]

def calculate_roi(outcome_prices, volume):
    """
    Calculate potential ROI for a market.
    outcome_prices: list of price strings like ["0.054", "0.946"]
    
    RVR 2.5x + ROC 10% strategy:
    - Risk/Reward Ratio of at least 2.5x
    - Return on Capital of at least 10%
    """
    if not outcome_prices or len(outcome_prices) < 2:
        return None, None
    
    try:
        prices = [float(p) for p in outcome_prices]
        if len(prices) != 2:
            return None, None
        
        # For binary markets, analyze both sides
        # If price is 0.3, potential gain is 0.7 (if it resolves to Yes)
        # RVR = potential_gain / risk = (1 - price) / price
        
        results = []
        for i, price in enumerate(prices):
            if price <= 0 or price >= 1:
                continue
            
            potential_gain = 1 - price
            risk = price
            rvr = potential_gain / risk if risk > 0 else 0
            roc = (potential_gain / price) if price > 0 else 0
            
            # Strategy: RVR >= 2.5 AND ROC >= 0.1 (10%)
            meets_criteria = rvr >= 2.5 and roc >= 0.1
            
            results.append({
                'side': i,
                'price': price,
                'rvr': rvr,
                'roc': roc,
                'meets_criteria': meets_criteria
            })
        
        return results
    except (ValueError, TypeError):
        return None

# Analyze markets
category_stats = defaultdict(lambda: {
    'total': 0,
    'total_volume': 0,
    'markets': []
})

print("Analyzing markets...")
for market in markets:
    question = market.get('question', '')
    description = market.get('description', '')
    outcome_prices = json.loads(market.get('outcomePrices', '[]'))
    volume = market.get('volumeNum', 0)
    
    category = categorize_market(question, description)
    
    # Calculate strategy fit
    roi_analysis = calculate_roi(outcome_prices, volume)
    
    market_data = {
        'id': market.get('id'),
        'question': question,
        'volume': volume,
        'outcome_prices': outcome_prices,
        'roi_analysis': roi_analysis,
        'meets_criteria': False
    }
    
    if roi_analysis:
        for analysis in roi_analysis:
            if analysis.get('meets_criteria'):
                market_data['meets_criteria'] = True
                break
    
    category_stats[category]['total'] += 1
    category_stats[category]['total_volume'] += volume
    category_stats[category]['markets'].append(market_data)

# Generate report
print("\n" + "="*80)
print("BACKTEST RESULTS BY CATEGORY")
print("="*80)

# Sort categories by number of markets
sorted_categories = sorted(category_stats.items(), key=lambda x: x[1]['total'], reverse=True)

category_summary = []
for category, stats in sorted_categories:
    total_markets = stats['total']
    total_volume = stats['total_volume']
    markets_meeting_criteria = sum(1 for m in stats['markets'] if m['meets_criteria'])
    
    pct_meeting_criteria = (markets_meeting_criteria / total_markets * 100) if total_markets > 0 else 0
    
    category_summary.append({
        'category': category,
        'total_markets': total_markets,
        'total_volume': total_volume,
        'markets_meeting_criteria': markets_meeting_criteria,
        'pct_meeting_criteria': pct_meeting_criteria
    })
    
    print(f"\n{category.upper().replace('_', '/')}")
    print(f"  Total Markets: {total_markets}")
    print(f"  Total Volume: ${total_volume:,.2f}")
    print(f"  Markets Meeting Criteria (RVR≥2.5x, ROC≥10%): {markets_meeting_criteria} ({pct_meeting_criteria:.1f}%)")
    
    # Show top markets meeting criteria
    qualifying_markets = [m for m in stats['markets'] if m['meets_criteria']]
    if qualifying_markets:
        # Sort by volume
        top_markets = sorted(qualifying_markets, key=lambda x: x['volume'], reverse=True)[:3]
        print(f"\n  Top Qualifying Markets:")
        for m in top_markets:
            print(f"    - {m['question'][:80]}")
            print(f"      Volume: ${m['volume']:,.2f}, Prices: {m['outcome_prices']}")

# Create markdown report
print("\n" + "="*80)
print("GENERATING BACKTEST_CATEGORIES.md")
print("="*80)

markdown_content = """# Market Category Backtest Analysis

## Strategy Parameters
- **Risk/Reward Ratio (RVR):** ≥2.5x
- **Return on Capital (ROC):** ≥10%
- **Markets Analyzed:** {}

## Category Rankings

### By Total Volume (Market Liquidity)
{}

### By Strategy Opportunity (% Markets Meeting Criteria)
{}

## Detailed Category Analysis

{}

## Key Findings

### Most Predictable Categories
{}

### Least Predictable Categories  
{}

## Recommendations

Based on the backtest analysis:

1. **Best Categories for Our Edge:**
{}

2. **Categories to Avoid:**
{}

3. **Volume Considerations:**
   - Higher volume markets generally have tighter spreads
   - Our strategy (RVR 2.5x+) tends to find more opportunities in lower probability outcomes
   - Consider balancing edge vs. liquidity

## Methodology

1. **Categorization:** Markets categorized by keyword matching in question/description
2. **Strategy Application:** Binary markets analyzed for both outcomes
3. **Criteria:** Trade qualifies if RVR ≥ 2.5x AND ROC ≥ 10%
4. **Data Source:** Gamma API (Polymarket), {} active markets analyzed

---
*Generated: 2026-02-06*
*Note: This is a snapshot analysis of currently active markets. For true historical backtest, resolved markets data needed.*
""".format(
    len(markets),
    # By volume
    '\n'.join([f"{i+1}. **{cat['category'].upper().replace('_', '/')}**: ${cat['total_volume']:,.2f} ({cat['total_markets']} markets)" 
               for i, cat in enumerate(sorted(category_summary, key=lambda x: x['total_volume'], reverse=True))]),
    # By opportunity
    '\n'.join([f"{i+1}. **{cat['category'].upper().replace('_', '/')}**: {cat['pct_meeting_criteria']:.1f}% ({cat['markets_meeting_criteria']}/{cat['total_markets']} markets)" 
               for i, cat in enumerate(sorted(category_summary, key=lambda x: x['pct_meeting_criteria'], reverse=True))]),
    # Detailed analysis
    '\n\n'.join([
        f"### {cat['category'].upper().replace('_', '/')}\n"
        f"- **Total Markets:** {cat['total_markets']}\n"
        f"- **Total Volume:** ${cat['total_volume']:,.2f}\n"
        f"- **Qualifying Markets:** {cat['markets_meeting_criteria']} ({cat['pct_meeting_criteria']:.1f}%)\n"
        f"- **Average Volume per Market:** ${cat['total_volume']/cat['total_markets'] if cat['total_markets'] > 0 else 0:,.2f}\n"
        for cat in sorted(category_summary, key=lambda x: x['total_volume'], reverse=True)
    ]),
    # Most predictable
    '\n'.join([f"- **{cat['category'].upper().replace('_', '/')}**: {cat['pct_meeting_criteria']:.1f}% of markets meet criteria" 
               for cat in sorted(category_summary, key=lambda x: x['pct_meeting_criteria'], reverse=True)[:3]]),
    # Least predictable
    '\n'.join([f"- **{cat['category'].upper().replace('_', '/')}**: {cat['pct_meeting_criteria']:.1f}% of markets meet criteria" 
               for cat in sorted(category_summary, key=lambda x: x['pct_meeting_criteria'])[:3]]),
    # Best categories
    '\n'.join([f"   - **{cat['category'].upper().replace('_', '/')}**: Strong opportunity rate ({cat['pct_meeting_criteria']:.1f}%) with ${cat['total_volume']:,.2f} volume" 
               for cat in sorted(category_summary, key=lambda x: x['pct_meeting_criteria'], reverse=True)[:3]]),
    # Worst categories
    '\n'.join([f"   - **{cat['category'].upper().replace('_', '/')}**: Low opportunity rate ({cat['pct_meeting_criteria']:.1f}%)" 
               for cat in sorted(category_summary, key=lambda x: x['pct_meeting_criteria'])[:3]]),
    len(markets)
)

with open('BACKTEST_CATEGORIES.md', 'w', encoding='utf-8') as f:
    f.write(markdown_content)

print("\n✅ BACKTEST_CATEGORIES.md created successfully!")
print(f"\nAnalyzed {len(markets)} active markets across {len(category_stats)} categories")
